### PR TITLE
ComboGeist - randomly juxtapose two notes

### PR DESCRIPTION
**Pull request by bfollington**
Before trying to port my genetic algorithm over I figured it would be smart to take some baby steps. 

The intention of this PR is to introduce the `ComboGeist`, which randomly picks two entries and presents them together in the feed (screenshot attached).

I want to add a `Synthesize` action that would create a new entry _pre-populated_ with the `/slashlinks` to both the entries shown in the feed. I'm not quite sure how to do this as the current flow requires the user to enter a slug for the entry before reaching the editor. That's not a problem in-and-of-itself but we'd need to thread some context through to support the pre-populated editor content.

Keen to hear any suggestions for how to achieve this 😁 

- [x] Register geist
- [x] Display combination of entries
- [x] Synthesize new entry from combination
- [x] Finalise naming scheme

### CODE REVIEW

**Comment by gordonbrander** [comment]
Amazing! This is exactly how I've been thinking about Geists as well.

There are two main CTAs I think Geists will be able to choose from:

- Open: opens the the note that is the focus of attention for editing
- Create (called synthesize here): creates a new note with a contents specified by the Geist

The distinction between open/create is actually looser than it might appear, since everything in Subconscious operates along a search-or-create flow. So a note will be either re-opened, or created depending upon whether it exists already or not.

Action `DetailAction.loadAndPresentDetail(slug:fallback:autofocus:)` will take care of looking up a note by slug, creating it if it doesn't exist already using `fallback` content, and then open up the editor, optionally autofocusing it. We dispatch this action down to detail from `FeedAction.openStory(EntryLink)` for the open action. https://github.com/subconsciousnetwork/subconscious/blob/main/xcode/Subconscious/Shared/Components/Feed.swift#L189.

We can use the same action for the create action. However, right now entries use the title of the note as fallback content if the note does not exist. For Geists to generate content, we probably want to have the option to include title AND further fallback content. I picture changing the action signature to something like `FeedAction.openStory(link: EntryLink, fallback: String)`.

This might also require a little bit of re-plumbing detail to support both EntryLink and fallback instead of just Slug and fallback.

@bfollington Let me know if you want to find some time on Hangouts to pair program on this. I anticipate it will be a minor change, but may touch a few files.

File: xcode/Subconscious/Shared/Services/ComboGeist.swift
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Services/ComboGeist.swift:31]: FYI, this form exists to allow us to easily load a JSON Tracery grammar from the app package. If we don't want to include Tracery-generated prompts with the two transcludes, then we don't need it.

However, I would lean toward including a small (possibly Tracery-generated) prompt with the story, just to frame up what you're supposed to do with these two juxtaposed stories.

Possible prompts:

- How might these be connected?
- What similarities can I see between these? Differences? 
- Can I combine these to create something new?
- Can I find three ways in which these might be connected?
- ...other good questions?

**Comment by gordonbrander** [comment]
@bfollington I put together a PR that updates the relevant actions to allow for custom fallback content to be specified https://github.com/subconsciousnetwork/subconscious/pull/341

After merging this PR, `StoryView`'s `action` parameter will support fallback content. Action will dispatch a `DetailAction.loadAndPresentDetail(link:fallback:autofocus:)` which will either open or create a note with custom fallback content, as necessary.

I'm going to hold off merging it into main just yet, since I have to cut a tag for the TestFlight release shortly, but I can merge it into main it shortly after we do the test flight release this coming week. In the mean time, feel free to merge it into your branch.

File: xcode/Subconscious/Shared/Services/ComboGeist.swift
Comment (bfollington) [reply xcode/Subconscious/Shared/Services/ComboGeist.swift:31]: Ah, that makes sense! It did feel like the structure was evaporating once I removed the grammar dependency. I agree that a small prompt is a nice touch, I'll re-introduce tracery and wire that up.

**Comment by bfollington** [comment]
@gordonbrander I rebased on your WIP PR and introduced some basic prompts that appear above the two entries. Thanks for the clarifications, I have a much better understanding of how you're thinking about actions for Geists now.

I said as much on Discord but I'd be keen to pair on getting the rest working, I think I could stumble my way there alone but programming is better together!

File: xcode/Subconscious/Shared/Components/Common/StoryComboView.swift
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Components/Common/StoryComboView.swift:62]: Presumably, we'll synthesize a unique link for the new content?

If so, my sense is that "Create" would be a good CTA here, even though it does the same thing under the hood.

Comment (bfollington) [reply xcode/Subconscious/Shared/Components/Common/StoryComboView.swift:62]: Yep! My intention was to create a new link (directly at the action dispatch site? Might make sense as shared utility function later?). I wasn't sure what label you'd prefer so I reverted to "Open" and I left it in a "draw the rest of the owl" state.

I'll push things along when I find a few spare hours this week, and/or if we line up a pairing session we'll be able to smash it out then 🙂

**Comment by gordonbrander** [comment]
I have merged the changes to `DetailAction.loadAndPresentDetail` into main.

**Comment by bfollington** [comment]
@gordonbrander I'm going to try and pick this up on Friday morning (Australia time) but if you have the time / inclination in the meantime feel free to branch off and move things forward.

**Comment by bfollington** [comment]
I took a stab at a naming scheme and format for the fallback content. Any feedback @gordonbrander?

_Moving this from draft to ready for review._

Screen Shot 2022-09-29 at 9 05 43 pm
Screen Shot 2022-09-29 at 9 05 57 pm

**Comment by gordonbrander** [comment]
Geist -> editor flow is looking good.

Something Chris and I have been talking about... When we build the block-based editor, the approach we have planned for transcludes looks like:

- `[[wikilinks]]` are rendered as text links, no transclude
- Bare `/slashlink` or `@bfollington/slashlink` are rendered as a transclude beneath the text block in which it appears.

The rationale is that wikilinks already have a textual representation, whereas slashlinks are more like a pointer.

So in future, the block containing `/subconscious /combogeist` would have rendered underneath it a transclude for `/subconscious`, followed by a transclude for `/combogeist`, in order. I think this is perfect!

File: xcode/Subconscious/Shared/Components/Common/StoryComboView.swift
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Components/Common/StoryComboView.swift:16]: We should probably come up with a convention for Geist-generated content like this. A couple ideas on possible conventions, without anchoring on any specific one:

- I like what you're doing here with `/combo/slug/slug`. OTOH, by default I lean toward flat naming schemes, since they create more chance of name collision (good, since it causes you to return to old notes and add to them https://subconscious.substack.com/p/knowledge-gardening-is-recursive).
- I like the "A-B" conjoining you're doing here. It makes sense that if we already created a combo, we might want to see that content instead of generating new content? How do we feel about `slug-slug` rather than `combo/slug/slug`?
- For Geist-generated content where we want a low chance of collision, maybe we default to `YYYY-MM-DD-name`? Not sure we need that here, though.

Comment (bfollington) [reply xcode/Subconscious/Shared/Components/Common/StoryComboView.swift:16]: I feel similarly torn. Nesting under `/combo` avoids polluting the root namespace and helps trace the origin of the entry, but the argument of maximising name collision is a strong one. 

If tracing the origin is actually useful we could also include `source: [[ComboGeist]]` in the content instead of  prefixing with `/combo`? I can see good arguments for `A-B` naming but longer slugs might be difficult to visually parse this way.

Also occurs to me that the same two entries (A and B) could be combined in A-B order or B-A order and I feel intuitively that those should point to the same combined slug? Perhaps instead of blindly concatenating A and B we should sort them them alphanumerically by slug beforehand (or maybe creation date?) I'll push a commit to try this out.

Having a defined ordering on slugs exposes some additional meaning from the `/combo/entry-a/entry-b` format, since viewing `/combo/entry-a` would list all the entries where you've combined `entry-a` with something else. 

e.g.
image

What's our orienting principle here? If we're maximising naming collisions then I vote to drop `combo/` and flatten to `slug-slug` but if we're maximising explorability I would vote to keep it.

Comment (gordonbrander) [reply xcode/Subconscious/Shared/Components/Common/StoryComboView.swift]: @bfollington let's go with the following:

- Order slugs by alpha
- combine as `slug-slug` (using `Slug(formatting: "\(a) \(b)")` should do the trick)

File: xcode/Subconscious/Shared/Components/Common/StoryComboView.swift
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Components/Common/StoryComboView.swift:34]: Nit: formatting

```swift
guard let slug = Slug(formatting: "\(first) \(second)") else {
    return
}
```

Not blocking merge though.

File: xcode/Subconscious/Shared/Components/Common/StoryComboView.swift
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Components/Common/StoryComboView.swift:43]: Note: I hardcoded these a while back just to get something on screen. Before we toggle Geists on we'll want to have some kind of dynamic header... probably the Geist's name.

This is just an FYI. No need to change.

File: xcode/Subconscious/Shared/Data/Geists/zettelkasten.json
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Data/Geists/zettelkasten.json:24]: @bfollington let's create a separate `combo.json` grammar for the combo geist.

File: xcode/Subconscious/Shared/SubconsciousApp.swift
Comment (gordonbrander) [inline xcode/Subconscious/Shared/SubconsciousApp.swift:681]: Note: in future, we should probably not force unwrap, and instead only register geist if it returns some value.

FYI only, no need to change.

**Review by gordonbrander — CHANGES_REQUESTED** [review]
Looks great. Just a couple of minor changes requested.

File: xcode/Subconscious/Shared/Data/Geists/zettelkasten.json
Comment (bfollington) [reply xcode/Subconscious/Shared/Data/Geists/zettelkasten.json]: Done!

File: xcode/Subconscious/Shared/SubconsciousApp.swift
Comment (bfollington) [reply xcode/Subconscious/Shared/SubconsciousApp.swift]: I took a swing at this, lmk what you think. I'm still learning the languages features of swift so I am likely ignorant of many patterns.

File: xcode/Subconscious/Shared/Components/Common/StoryComboView.swift
Comment (bfollington) [reply xcode/Subconscious/Shared/Components/Common/StoryComboView.swift]: Fixed

File: xcode/Subconscious/Shared/Data/Geists/combo.json
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Data/Geists/combo.json:7]: I love this prompt @bfollington

File: xcode/Subconscious/Shared/SubconsciousApp.swift
Comment (gordonbrander) [inline xcode/Subconscious/Shared/SubconsciousApp.swift:672]: TY!

**Review by gordonbrander — APPROVED** [review]
Let's goooooooooooooooo!!! 🚀🌔

**Pull request merged**
